module: add createRequire method

MylesBorins opened the pull request:
This is an abstraction on top of creatRequireFromPath that can accept
both paths, URL Strings, and URL Objects.


<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

benjamingr commented on test/parallel/test-module-create-require.js:
```diff
@@ -5,8 +5,17 @@ const assert = require('assert');
 const path = require('path');
 
 const { createRequireFromPath } = require('module');
+const { createRequire } = require('module');
```
Maybe add a test for what happens with invalid input? (like a 'junk' string as URL)

MylesBorins replied:
@benjamingr a garbage string is not "invalid input" afaict based on the implementation of `createRequireFromPath`. Things would fail later when using the generated `require`, but this behavior is specific to Module and not createRequire. I personally don't think we need extra tests TBQH

MylesBorins replied:
more tests added

mscdex commented:
Is there a reason this can't be a part of `createRequireFromPath()` instead of creating a separate function?

MylesBorins commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -828,6 +828,20 @@ Module.runMain = function() {
+    path = file;
+  }
+
+  return Module.createRequireFromPath(path);
```
fixed

MylesBorins commented:
Updated code based on feedback from @devsnek. Have also added documentation to `modules.md`.

@mscdex I don't think this makes sense to add to `createRequireFromPath` as that API explicitly states it is meant for paths. I think it might make sense to deprecate createRequireFromPath and make it an internal API, as `createRequire` is far more useful IMHO.

MylesBorins commented:
CI is actually green the fail above is a lie. If there are no objections to this PR I plan to land it as is on Monday

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -828,11 +828,27 @@ Module.runMain = function() {
+  try {
+    path = fileURLToPath(filename);
+  } catch (e) {
+    if (e.code === 'ERR_INVALID_URL_SCHEME') {
```
`fileURLToPath` could throw a `ERR_INVALID_FILE_URL_PATH` error here, which seems like it would then cause an object type error down the line given the fallthrough, which seems non-ideal.

I do think we might need to be a bit stricter here.

How about a logic like:

If `filename` is a URL object or file: string (`typeof filename === 'object' && filename instanceof URL || typeof filename === 'string' && filename.startsWith('file:')`), then pass it to `fileURLToPath` before the `createRequireFromPath` function. Any error can then be fully propagated without a try-catch guard.

MylesBorins replied:
This logic didn't exactly work because it is possible to have url type strings e.g. `'html:` that will pass through and then not fail early enough.

I've updated this PR to require absolute paths... so the logic here was simplified a bit

MylesBorins commented on doc/api/modules.md:
````diff
@@ -905,6 +905,23 @@ by the [module wrapper][]. To access it, require the `Module` module:
+requireUtil('./some-tool');
+```
+
 ### module.createRequireFromPath(filename)
````
Done

MylesBorins commented on test/parallel/test-module-create-require.js:
```diff
@@ -5,8 +5,17 @@ const assert = require('assert');
+const reqToo = createRequire(u);
+assert.deepStrictEqual(reqToo('./experimental'), { ofLife: 42 });
+
+assert.throws(() => { createRequire('https://github.com/nodejs/node/pull/27405/'); }, {
```
Fixed

MylesBorins commented:
Changed a few things here:

* createRequire now only supports absolute paths (either path or URL)
  - this is because the underlying logic in Module for relative paths is undefined / confusing e.g. caching cwd(). This limitation, which we could choose to remove later, guards against unexpected behavior (which was present in our documentation)
* createRequireFromPath is now docs deprecated
  - please LMK if I need to do anything more, haven't deprecated before

PTAL

MylesBorins commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -824,12 +825,30 @@ Module.runMain = function() {
+  if (typeof filename === 'string' && filename.startsWith('.')) {
+    throw new ERR_INVALID_CREATE_REQUIRE_PATH(filename);
+  } else if (typeof filename === 'object' && filename instanceof URL ||
+    typeof filename === 'string' && !filename.startsWith('/')) {
```
@guybedford as we now enforce absolute paths this logic should cause any string based URLs to go through `fileURLToPath` and fail accordingly

SMotaal replied:
This seems to align a lot better in concept to resolve the odd behaviour I was noting earlier — thanks 👍

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -824,12 +825,30 @@ Module.runMain = function() {
+  if (typeof filename === 'string' && filename.startsWith('.')) {
+    throw new ERR_INVALID_CREATE_REQUIRE_PATH(filename);
+  } else if (typeof filename === 'object' && filename instanceof URL ||
+    typeof filename === 'string' && !filename.startsWith('/')) {
```
`filename.startsWith('/')` likely won't work on Windows? `path.isAbsolute` may be worth using here as it explicitly does the Windows drive letter detection, distinguishing drive letters from file URLs.

Alternatively we could do the inverse - `filename.startsWith('file:')`

MylesBorins commented:
One more stab at it

Differences:

* now fails with non-url objects early
* swallows all errors from `fileURLToPath` and has a single unified error
* explicitly only support
  - file URL objects
  - file URL string
  - absolute path string

I believe this has scoped the use of this API enough that it guards against most footguns PTAL

BridgeAR commented:
Using `gzemnid` I get the following hits using this API currently:

```
chiml-0.6.3.tgz/src/index.ts:14:const { createRequireFromPath } = require("module");
chiml-0.6.3.tgz/src/index.ts:38:        const relativeRequire = createRequireFromPath(containerFile);
pobpack-utils-7.3.2.tgz/src/webpack/createModuleConfig.ts:7:// import { createRequireFromPath } from 'module';
pobpack-utils-7.3.2.tgz/src/webpack/createModuleConfig.ts:8:// const requireFromPwd = createRequireFromPath(process.cwd());
bpkg-0.2.1.tgz/lib/resolver.js:348:    // Note that module.createRequireFromPath
carmi-1.0.143.tgz/src/analyze-dependencies.js:76:  // const {createRequireFromPath} = require('module')
carmi-1.0.143.tgz/src/analyze-dependencies.js:77:  // const requireUtil = createRequireFromPath(modulePath)
get-app-root-path-0.0.11.tgz/index.js:254:  function createRequireFromPath(filename) {
get-app-root-path-0.0.11.tgz/index.js:288:        customRequire = (cjs.createRequireFromPath || createRequireFromPath)(fname)
parcel-plugin-ttypescript-1.0.1.tgz/TTypescriptAsset.js:2:const { createRequireFromPath } = require('module');
parcel-plugin-ttypescript-1.0.1.tgz/TTypescriptAsset.js:10:const _require = createRequireFromPath(tsAssetPath);
require-at-1.0.4.tgz/require-at.js:9:const createRequireFromPath =
require-at-1.0.4.tgz/require-at.js:10:  Module.createRequireFromPath ||
require-at-1.0.4.tgz/require-at.js:63:      xRequire = createRequireFromPath(Path.join(xdir, "._require-at_"), xdir);
require-from-path-1.0.4.tgz/require-from-path.js:9:const createRequireFromPath =
require-from-path-1.0.4.tgz/require-from-path.js:10:  Module.createRequireFromPath ||
require-from-path-1.0.4.tgz/require-from-path.js:63:      xRequire = createRequireFromPath(Path.join(xdir, "._require-from-path_"), xdir);
```

We should probably open PRs to fix `carmi` and `require-at` since those two are the most used modules from the list (because of the deprecation).

MylesBorins commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -824,12 +825,33 @@ Module.runMain = function() {
+    } catch {
+      throw new ERR_INVALID_CREATE_REQUIRE_ARGUMENT(filename);
+    }
+  } else {
```
nope, the API also support URL objects

BridgeAR replied:
I mean in this code branch. The URL objects where already handled at this point.

MylesBorins replied:
ahhh, done PTAL

addaleax commented on doc/api/modules.md:
```diff
@@ -905,15 +905,35 @@ by the [module wrapper][]. To access it, require the `Module` module:
+-->
+
+* `filename` {string|URL} Filename to be used to construct the require
+  function. Must be an absolute path.
```
Can we call this `path` in the documentation if it doesn’t have to refer to an actual file? (That’s also an issue that `createRequireFromPath()` has, I guess.)

MylesBorins replied:
I updated the docs to be more specific, does that work better? If not please feel lmk what else to change or feel free to push a commit

addaleax replied:
I was thinking about the parameter name itself – `filename` doesn’t seem accurate if it can also be a directory (which too is a kind of file, but that’s not always intuitive), and if we don’t currently enforce it (aka add a `stat` call). `path` would be plainer and seems more accurate to me.

MylesBorins replied:
I can change it to `filepath`, at least that would be more accurate... thoughts?

addaleax replied:
I guess? My issue was more with the `file` bit than the `name` bit :smile: But it’s also really not that important

MylesBorins replied:
I'm struggling figuring out the best name. The argument can be a file URL string, a file URL object, and an absolute path string. I'm concerned path is not inclusive of URL, but maybe I'm over thinking it.

sigh

MylesBorins commented:
I've updated based on all the comments PTAL

BridgeAR commented:
It still needs an entry in `doc/api/deprecations.md`. Besides that it's LGTM.

MylesBorins commented:
@BridgeAR documented deprecation PTAL

benjamingr commented:
changes still lgtm

Trott commented on doc/api/deprecations.md:
```diff
@@ -2439,6 +2439,19 @@ The `_channel` property of child process objects returned by `spawn()` and
 instead.
 
+<a id="DEP0130"></a>
+### DEP0130: Module.createRequireFromPath()
```
Should this be `DEP00XX` or something like that? https://github.com/nodejs/node/blob/be6b1a2b3a5e01491b829ac9a5ed7602807af2de/doc/releases.md#step-3-update-any-replaceme-and-dep00xx-tags-in-the-docs

MylesBorins replied:
fixed

MylesBorins commented:
latest CI was green!

landed in 411063c6f5
